### PR TITLE
Support static library build.

### PR DESCRIPTION
**Pull request by mcmtroffaes**
Allows users to build the library statically, by passing ``-DBUILD_SHARED_LIBS=OFF`` on the command line.

### CODE REVIEW

**Comment by qyot27** [comment]
Out of curiosity, what's going to be linking it statically?

But even then, https://github.com/AviSynth/AviSynthPlus/issues/130 is still something that likely needs addressing, and libraries typically have differences between shared and static configurations (similar issues are currently handled in [avs/capi.h](https://github.com/AviSynth/AviSynthPlus/blob/master/avs_core/include/avs/capi.h), but whether this is adequate for building as static is another question).

**Comment by mcmtroffaes** [comment]
That's a good question. I was hoping to see if avisynthlib could be statically linked into ffmpeg on windows to avoid having to bundle a dll. There might be some issues to iron out, true. This patch is sufficient for the static build to succeed on Windows (without plugins), but I haven't tested it in other applications or on other platforms. For the record, ffmpeg currently dynamically loads avisynth, presumably to allow alternative implementations to be dropped in.

If a static build isn't a configuration that the avisynthplus project wants to support, feel free to close this.

**Comment by pylorak** [comment]
Some further insight into the consequences of linking statically to Avisynth.

**TL;DR:** Linking statically to Avisynth changes the license of the built binary, so be warned!

**Long version:**

> For the record, ffmpeg currently dynamically loads avisynth, presumably to allow alternative implementations to be dropped in.

There is actually a second reason for that, and that's, very unfortunately, Avisynth's licence. Unfortunate because Avisynth will never be able to change it. It has already tried, very aggressively even, but the task was impossible in the end.

Avisynth only allows linking to non-GPL software if 1) only the API in avisynth.h is used (this is easily satisfied), and 2) if the other software is an "independent module".  The problem is in the definition of what an "independent module" is, which is somewhat unclear, but in the safe interpretation, one requirement is that the link between the two projects must not be static.

So mind you, if you link Avisynth statically into ffmpeg, then your ffmpeg build becomes GPL, and then all users of your ffmpeg also become GPL. This can be problematic, because many projects want to (and are only allowed to) link to an **L**GPL version of ffmpeg.

There is a solution to this problem: Only allow statically linking Avisynth in ffmpeg if ffmpeg is built both without "--enable-gpl" and without "--enable-nonfree". For this you'd need to send in patches to the ffmpeg project too. This is only a solution in your case though. Most other projects don't have a GPL/LPGL mode switch like ffmpeg does. Due to the already existing wide ecosystem that Avisynth has (where all users have for the better or worse learned that can they can simply link to Avisynth without taking precautions), I suggest that building Avisynth as a static libarary should produce a big fat build warning that the resulting binary will only allow GPL users.

**Comment by mcmtroffaes** [comment]
Thanks for your detailed response.

I agree, licensing is a very important point. It's the vcpkg build that I'm mainly interested in, and at least in vcpkg, FFmpeg already carries a GPL license when compiled with AviSynthPlus support (statically or dynamically), since using the AviSynthPlus dll together with FFmpeg forces the combined result to be GPL (if distributed together), according to the free software foundation (see https://www.gnu.org/licenses/gpl-faq.en.html#GPLStaticVsDynamic). As per that interpretation, ffmpeg should always be GPL when avisynth is enabled. But you're right, it currently is not, presumably due to a different reading of the GPL.

Anyway, I'm just trying to see if a static build of ffmpeg, with avisynthplus, in vcpkg, under a GPL license, could be achieved. If there's reasons not to do that, I'll respect that.

**Comment by mcmtroffaes** [comment]
Correction: ffmpeg automatically enables GPL when built with avisynth support, so ffmpeg already does the right thing (I just tested it).

**Comment by qyot27** [comment]
>to avoid having to bundle a dll
>For the record, ffmpeg currently dynamically loads avisynth, presumably to allow alternative implementations to be dropped in.

It ties into the second point there, in that the reason for the dynamic loader was actually to support multiple point versions of AviSynth at the same time without having to rebuild FFmpeg.  At the time the new demuxer was written (2012-early 2013; the old AviSynth 'demuxer' in libavformat was basically just a Video for Windows input source), AviSynth+ didn't exist yet, and Windows users weren't unified in whether to use the 2.6 alphas or stick with 2.5.8 (and x264 already had set a precedent of dynamically loading AviSynth.dll).  A little of that has remained, in that keeping 2.6 support around doesn't really harm anything, but FFmpeg did drop support for 2.5 on Windows several years ago (the last vestiges of the 2.5 API to leave was in April when AvxSynth support¹ was switched out with AviSynth+ on non-Windows).

¹which, of course, is why the new demuxer exists; it was initially written by the AvxSynth team so that there was a usable cross-platform demuxer.

The major reason for it in both x264 and FFmpeg, though, is the basic nature of what dynamic loading allows for in contrast to static or shared linking: if AviSynth(+) isn't installed, the client won't crash, but if the user then installs AviSynth(+), the client can find and use it without the user having to rebuild.  Package management usually assumes a client program is going to link - dynamically or statically - to a library, so the -dev package (or equivalent) typically ships both the shared/static library and the headers.  But a -dev package for AviSynth+ currently only **needs** the headers, so that FFmpeg can be built with support for it, and then that same FFmpeg binary can be distributed to users who want AviSynth support and already have it installed (or choose to it install it later), as well as those that don't care and don't have it installed.

Allowing a static (or even a regular shared) link of AviSynth into FFmpeg or x264 will require sequestering the dynamic loading pieces behind ifdefs and defining the abstraction away so that the API calls throughout the demuxer don't get needlessly duplicated for each use case.

And yes, the licensing issue isn't really a problem for FFmpeg; it's required `--enable-gpl` in order to use `--enable-avisynth` for a few years now (the [GPL linking exception](https://en.wikipedia.org/wiki/GPL_linking_exception) in `avisynth_c.h` apparently wasn't convincing).

>Due to the already existing wide ecosystem that Avisynth has (where all users have for the better or worse learned that can they can simply link to Avisynth without taking precautions), I suggest that building Avisynth as a static libarary should produce a big fat build warning that the resulting binary will only allow GPL users. EDIT: Hence I suggest this patch only be accepted if it is supplemented with this warning.

Agreed.

Broadly, I'm not opposed to making it possible to build as a static library, as even though it's highly likely that FFmpeg and x264 will stick to dynamic loading, it doesn't mean there aren't/won't be scenarios where a statically linked form wouldn't be useful.  We just need to make sure that static builds actually function correctly.  I'd argue that as long as the static library can still load external plugins, then it's fine (notwithstanding what I mention below about musl).  Otherwise we'd be kneecapping ourselves by disallowing any plugin we don't have source control of.

Particularly, a scenario where regular linking would be beneficial is when dealing with musl: musl does not support dlopen, so on any Linux distro that uses musl as its CRT rather than glibc, AviSynth+ won't work (I ran head-first into this when I tested Adélie Linux, and only fully realized it was a dlopen incompatibility when the musl version of Void-ppc acted the exact same way).  It also makes a bit more sense when you also factor in that we provide a .pc file for pkg-config...but nothing uses it, either (only some checkinstall instructions that grab the version number from it).

**Comment by mcmtroffaes** [comment]
Thanks for the detailed explanation, and I fully agree with the points raised. For now, I've added a commit which emits a GPL warning during the cmake configure phase when a static library is built. However, I understand completely if you were to close this PR until it is clear that regular linking is a scenario you want to fully support as upstream, if ever. I appreciate that this is non-trivial, even if there are some valid use cases.

**Comment by pylorak** [comment]
Oh, I didn't realize ffmpeg already requires --enable-gpl, I just assumed it doesn't, because there's no reason for it (as long as Avisynth is linked in dynamically). Either they got scared of the linking exception in the licence, or they just didn't look hard enough (e.g. they saw the GPL header and stopped reading).

Thanks for taking my suggestion with the warning (thumbs up!).

> 
> 
> However, I understand completely if you were to close this PR until it is clear that regular linking is a scenario you want to fully support as upstream, if ever.

Nobody wants to close this PR. I'm sure once it's been verified that a static Avs+ technically works similarly well as a dynamic one (for example external plugins are loadable), the PR will be accepted immediately.

**Comment by mcmtroffaes** [comment]
Sure, thanks!

**Comment by qyot27** [comment]
In one example of something that may be a good testbed for something like this, in #193 I suggest that we have our own upstream piping tool.  A static build of the AviSynth+ core linked into such a tool which is also provided by us does make sense (not to mention that NUT would be an improvement over the y4m piping tools).

**Comment by qyot27** [comment]
It took awhile, but I finally found time to test this, and it does appear that yes, a static build of libavisynth *can* still use its internal plugin loader to dlopen external plugins.  I tested it against a FFmpeg [where I disabled all the dynamic loading bits in libavformat/avisynth.c](https://github.com/qyot27/FFmpeg/commit/e70f70a6d4e096bdf146bbf4a444299dc35421ef)¹ (and even went so far as to rename the libavisynth.so* files in `/usr/local/lib` so it couldn't find them either), and it could handle both `Version()` and an FFMS2-loaded video and a bit of filtering with `AddGrainC` (which was also recently made to work on Linux).

¹initial rough draft, I did a dumb Find+Replace on the `avs_library.` bit, so it even hit some that were behind the `#if 0`s that would probably be changed eventually to something more descriptive.

At least in FFmpeg's case, it did expose the one thing missing from this small patchset: avisynth.pc needs to have the correct libraries added to `Libs.private` so that pkg-config can correctly populate the link list.  On Linux, and with GCC, that seems like it would be just
```
Libs.private: -lm -lstdc++ -ldl
```
but I'm not sure how much that might need to change on a different platform, like Haiku, or when using Clang instead of GCC.

So I'd say it's safe to consider this to be in the pipeline to show up in 3.7, as anything that wants to use a static build will need to be able to query pkg-config for it, and letting this in as part of the bump to 3.7 makes that easier.

**Comment by mcmtroffaes** [comment]
Great! Thanks for the work, and for the update.

**Comment by qyot27** [comment]
Merged in as part of 3.7.
